### PR TITLE
AudioKit setInputDevice bug fixes [iOS]

### PR DESCRIPTION
**Pull request by bakeddean**
Hopefully fixes issues like #808, #1706, #1532, where users (including myself), cannot change the input device.

- Refactor the AudioKit.setInputDevice(_ input: AKDevice) method. Because of differences in the way AudioKit.inputDevice &  AudioKit.inputDevices instantiate their AKDevice, String comparisons like ```dev.uid == input.deviceID``` are false, and fail to set ```AVAudioSession.sharedInstance().setPreferredInput()```
eg "Built-In Microphone" == "Built-In Microphone Back"
The method also potentially sets the preferred device twice, and is generally a bit hard to follow. Also, let the errors propagate, at the moment they are swallowed for iOS but not mac.

- Added convenience initialiser to AKDevice to take a AVAudioSessionPortDescription. Allows easier instantiation of AKDevices in AudioKit.inputDevice & AudioKit.inputDevices, and is more consistent.

- Added isEqual in AKDevice to make it easier to compare devices.
eg ```cell.accessoryType = (inputDevice == currentInputDevice) ? .checkmark : .none```

- Added example of input device switching to the MicrophoneAnalysis example. Built the AudioKit framework from my changes and tested MicrophoneAnalysis and every seems to be working much better.

Cheers.

### CODE REVIEW

**Comment by aure** [comment]
I'll get this passing Travis and then I'll be happy to accept this awesomeness.

**Comment by bakeddean** [comment]
Thanks. Silly me, I didn't realise AVFoundation was iOS only.
I've thought about this overnight, and think this could be nicer still.
Have you done any work on this yet? I could push some more changes if not.

Ta.

**Comment by aure** [comment]
I haven't done anything yet, I'll let you give it a go.

**Comment by bakeddean** [comment]
Thanks.

**Comment by bakeddean** [comment]
Righto. I'm happier with this, see what you think.

**AKDevice**
- Added computed properties for portDescription & dataSource
- Fix the conditional compilation issues eg. #if !os(macOS)

**AudioKit**
- Simplify the setInputDevice() method further.

**AKError**
- Added DeviceNotFound error

**MicrophoneAnalysis**
- Add nav bar and moved Input button to said nav bar
- Improved landscape support
- Code simplification
- Added README & screen shots.

**Comment by bakeddean** [comment]
Hmm. Maybe portDescription & dataSource should be open vars's?

**Comment by aure** [comment]
Do you think you could address issues like this with similar changes, except on output? https://github.com/AudioKit/AudioKit/issues/1654

**Comment by bakeddean** [comment]
I don't have the hardware devices to test his exact problem at the moment. It might be when he sets the output, AudioKit potentially creates a route change. Of overrideAudioOutputPort, Apple says "This change remains in effect only until the current route changes"

I think your saying, don't worry about all this, and implement ```AudioKit.setOutputDevice```, which is not implemented yet on iOS. I was thinking about this last night. Its seems logical that iOS should support multiple output devices. Dongles like the iRig HD2 or even simple USB headset/mic devices plugged into the camera connection kit should present as output options.

At the moment I've only got input devices like the Sonuus i2m, but I was going to get a cheap dongle like this:
s-l225

Hopefully its class compliant.

And see if it comes up as an output device, and implement setOutputDevice and see how things go.

Cheers.

**Comment by aure** [comment]
I'd be happy to reimburse you for an inexpensive device that would allow you to delve further, just let me know.

**Comment by bakeddean** [comment]
I'll give these cheapies a go first, they are only a few dollars. iOS is pretty good with USB audio as long as its class compliant/driverless.
